// htmlgrabr/build/module/helpers.js
import * as mime from 'mime-types';
import { URL } from 'url';
const isValidUrl = (url) => {
    try {
        const u = new URL(url);
        return u.protocol === "http:" || u.protocol === "https:";
    }
    catch (_) {
        return false;
    }
};
/**
 * Returns true if the HTML node is an element.
 * @param node HTML node
 */
export const isElementNode = (node) => node.nodeType === 1;
/**
 * Extract base URL from headers of a DOM.
 * @param doc DOM to process
 * @returns the base URL
 */
export function extractBaseUrl(doc) {
    let result = null;
    if (!doc.head) {
        return result;
    }
    const $base = doc.head.getElementsByTagName('base')[0];
    if ($base && $base.hasAttribute('href')) {
        let baseUrl = $base.getAttribute('href') || '';
        if (/^\/\//i.test(baseUrl)) {
            baseUrl = 'http:' + baseUrl;
        }
        if (isValidUrl(baseUrl)) {
            result = baseUrl;
        }
    }
    return result;
}
/**
 * Extract Open Graph properties from headers of a DOM.
 * @param doc DOM to process
 * @returns Open Graph properties
 */
export function extractOpenGraphProps(doc) {
    const result = {};
    if (!doc.head) {
        return result;
    }
    const $metas = Array.from(doc.head.getElementsByTagName('meta'));
    for (const $meta of $metas) {
        const attr = $meta.getAttribute('property');
        if (attr !== null && attr.startsWith('og:')) {
            const prop = attr.substr(3);
            result[prop] = $meta.getAttribute('content');
        }
        if (!isValidUrl(result.url)) {
            result.url = null;
        }
    }
    return result;
}
/**
 * Extract all images from a DOM.
 * @param doc DOM to process
 * @param illustration if provided, the illustration is added to the result
 * @returns array of image meta data
 */
export function extractImages(doc, illustration) {
    const result = [];
    if (illustration) {
        // Add document Open Graph illustration in first position
        // This in order to be the document illustration.
        result.push({
            src: illustration,
            contentType: mime.lookup(illustration) || '',
        });
    }
    const $images = Array.from(doc.getElementsByTagName('img'));
    for (const $img of $images) {
        if ($img.hasAttribute('src')) {
            const src = $img.getAttribute('src');
            if (src && !/^data:/i.test(src)) {
                result.push({
                    src,
                    contentType: mime.lookup(src) || '',
                });
            }
        }
    }
    return result;
}
