// htmlgrabr/build/main/htmlgrabr.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsdom_1 = require("jsdom");
const readability_1 = require("@mozilla/readability");
const node_fetch_1 = __importStar(require("node-fetch"));
const pretty_1 = __importDefault(require("pretty"));
const url_1 = require("url");
const blocked_host_1 = require("./blocked-host");
const helpers_1 = require("./helpers");
const sanitize_1 = require("./sanitize");
const rules_1 = require("./rules");
const DefaultConfig = {
    debug: false,
    pretty: false,
    headers: new node_fetch_1.Headers({
        'User-Agent': 'Mozilla/5.0 (compatible; HTMLGrabr/1.0)',
    }),
    rules: rules_1.DefaultRules,
    isBlockedHost: blocked_host_1.isBlockedHost,
};
class HTMLGrabr {
    constructor(config = {}) {
        this.config = Object.assign(Object.assign({}, DefaultConfig), config);
    }
    /**
     * Grabs the content of a page from HTML content.
     * @param content a string that contains HTML code
     * @param baseURL a string that contains HTML base URL
     * @returns a page object
     */
    async grab(content, baseURLFallback) {
        const { debug, isBlockedHost, rewriteURL } = this.config;
        // Load content into a virtual DOM
        const dom = new jsdom_1.JSDOM(content, {
            url: baseURLFallback,
        });
        const doc = dom.window.document;
        // Apply rule if exists
        if (baseURLFallback) {
            const { hostname } = new url_1.URL(baseURLFallback);
            const redirect = this.applyRules(hostname.replace(/^(www\.)/, ""), doc);
            if (redirect) {
                return this.grabUrl(redirect);
            }
        }
        // Extract base URL
        const baseURL = helpers_1.extractBaseUrl(doc) || baseURLFallback;
        // Extract Open Graph properties
        const ogProps = helpers_1.extractOpenGraphProps(doc);
        // Extract images
        const images = helpers_1.extractImages(doc, ogProps.image);
        // Use Readability.js to extract HTML content
        const reader = new readability_1.Readability(doc, { debug });
        const article = reader.parse();
        if (debug) {
            console.log('article after Readability parsing:', article);
        }
        // Sanitize content
        let html = sanitize_1.sanitize(article.content, { baseURL, debug, isBlockedHost, rewriteURL });
        if (debug) {
            console.log('HTML content after sanitization:', html);
        }
        if (this.config.pretty) {
            html = pretty_1.default(html, { ocd: true });
        }
        return {
            title: ogProps.title || article.title,
            url: ogProps.url || baseURL,
            image: ogProps.image,
            html,
            text: article.textContent,
            excerpt: article.excerpt,
            length: article.length,
            images,
        };
    }
    /**
     * Grabs the content of a remote HTML page.
     * @param url the URL to fetch and process
     * @returns a page object
     */
    async grabUrl(url) {
        let req = new node_fetch_1.Request(url.toString(), {
            headers: this.config.headers,
            method: 'HEAD'
        });
        let res = await node_fetch_1.default(req);
        if (!res.ok) {
            throw new Error(`bad status response: ${res.statusText}`);
        }
        const contentType = res.headers.get('Content-Type');
        if (!contentType.startsWith('text/html')) {
            throw new Error(`unsupported content type: ${contentType}`);
        }
        req = new node_fetch_1.Request(url.toString(), {
            headers: this.config.headers,
        });
        res = await node_fetch_1.default(req);
        if (!res.ok) {
            throw new Error(`bad status response: ${res.statusText}`);
        }
        const body = await res.text();
        return this.grab(body, url.toString());
    }
    applyRules(hostname, doc) {
        if (!this.config.rules.has(hostname)) {
            return null;
        }
        const rule = this.config.rules.get(hostname);
        const node = doc.querySelector(rule.selector);
        if (node && helpers_1.isElementNode(node)) {
            if (rule.type === 'redirect' && (node.hasAttribute('src') || node.hasAttribute('href'))) {
                const src = node.getAttribute('src') || node.getAttribute('href');
                return new url_1.URL(src);
            }
            else {
                doc.body.childNodes.forEach((n) => n.remove());
                doc.body.prepend(node);
            }
        }
        return null;
    }
}
exports.HTMLGrabr = HTMLGrabr;
exports.default = HTMLGrabr;
